pgrx/src/datum/array.rs

```diff
@@ -97,107 +103,46 @@ impl<'a, T: FromDatum + serde::Serialize> serde::Serialize for Array<'a, T> {
-                array,
-                oid,
-                elem_layout.size.as_typlen().into(),
-                matches!(elem_layout.pass, PassBy::Value),
-                elem_layout.align.as_typalign(),
-                &mut elements,
-                &mut nulls,
-                &mut nelems,
-            )
-        };
-
-        let nelems = nelems as usize;
-
-        // Check our RawArray len impl for correctness.
-        assert_eq!(nelems, len);
-
-        // Detoasting the varlena may have allocated a fresh ArrayType.
-        // Check pointer equivalence, then pfree the palloc later if it's a new one.
-        let needs_pfree = ptr.as_ptr().cast() != array;
-        let mut raw = unsafe { RawArray::from_ptr(NonNull::new_unchecked(array)) };
-
+        let elem_layout = Layout::lookup_oid(oid);
+        let nelems = raw.len();
         let null_slice = raw
             .nulls_bitslice()
             .map(|nonnull| NullKind::Bits(unsafe { &*nonnull.as_ptr() }))
             .unwrap_or(NullKind::Strict(nelems));
+        let _datum_slice = OnceCell::new();
 
-        // The array was just deconstructed, which allocates twice: effectively [Datum] and [bool].
-        // But pgrx doesn't actually need [bool] if NullKind's handling of BitSlices is correct.
-        // So, assert correctness of the NullKind implementation and cleanup.
-        // SAFETY: The pointer we got should be correctly constructed for slice validity.
-        let pallocd_null_slice =
-            unsafe { PallocSlice::from_raw_parts(NonNull::new(nulls).unwrap(), nelems) };
         #[cfg(debug_assertions)]
-        for i in 0..nelems {
-            assert!(null_slice.get(i).unwrap().eq(unsafe { pallocd_null_slice.get_unchecked(i) }));
-        }
+        let Ok(()) = _datum_slice.set(unsafe {
+            let (datums, _bools) = raw.deconstruct(elem_layout);
+            PallocSlice::from_raw_parts(NonNull::new(datums).unwrap(), nelems)
+        }) else {
+            panic!("oh no, the debug code exploded!")
+        };
 
-        // SAFETY: This was just handed over as a palloc, so of course we can do this.
-        let datum_slice =
-            Some(unsafe { PallocSlice::from_raw_parts(NonNull::new(elements).unwrap(), nelems) });
-
-        Array {
-            needs_pfree,
-            raw: Some(raw),
-            nelems,
-            datum_slice,
-            null_slice,
-            elem_layout,
-            _marker: PhantomData,
-        }
+        Array { raw, nelems, _datum_slice, null_slice, elem_layout, _marker: PhantomData }
     }
 
-    pub fn into_array_type(mut self) -> *const pg_sys::ArrayType {
-        let ptr = mem::take(&mut self.raw).map(|raw| raw.into_ptr().as_ptr() as _);
-        mem::forget(self);
-        ptr.unwrap_or(ptr::null())
+    /// Rips out the underlying `pg_sys::ArrayType` pointer.
+    /// Note that Array may have caused Postgres to allocate to unbox the datum,
+    /// and this can hypothetically cause a memory leak if so.
```

Comment:
Yeah, that's the only thing I can think of. I can delete that remark if it's confusing!

Reply:
Maybe this is a little more subtle than we think?  If we give the pointer back to the user, it becomes their responsibility to free it (or to let it dangle until Postgres does garbage collection, which is probably the right answer), but there's no way for the user to know if it could safely be freed.

Could we just remove this function entirely?  I don't see a burning need to give out an ArrayType pointer, do you?

Reply:
One user uses this to go from Array to RawArray, I believe.